src/perf/event.cpp

```
@@ -0,0 +1,658 @@
+{
+    EventGuard proc_ev;
+    EventGuard sys_ev;
+
+    try
+    {
+        proc_ev = open(Thread(0));
+    }
+    catch (const std::system_error& e)
+    {
+    }
+
+    try
+    {
+        sys_ev = open(*supported_cpus().begin());
+    }
+    catch (const std::system_error& e)
+    {
+    }
+
+    if (proc_ev.get_fd() == -1)
+    {
+        if (sys_ev.get_fd() == -1)
+            availability_ = Availability::UNAVAILABLE;
+        else
+            availability_ = Availability::SYSTEM_MODE;
+    }
+    else
+    {
+        if (sys_ev.get_fd() == -1)
+            availability_ = Availability::PROCESS_MODE;
+        else
+            availability_ = Availability::UNIVERSAL;
+    }
+}
+
+bool Event::degrade_precision()
+{
+    /* reduce exactness of IP can help if the kernel does not support really exact events */
+    if (attr_.precise_ip == 0)
+    {
+        return false;
+    }
+    else
+    {
+        attr_.precise_ip--;
+        return true;
+    }
+}
+
+SysfsEvent::SysfsEvent(const std::string& ev_name, bool enable_on_exec) : Event()
+{
+    set_common_attrs(enable_on_exec);
+
+    // Parse event description //
+
+    /* Event description format:
+     *   Name of a Performance Monitoring Unit (PMU) and an event name,
+     *   separated by either '/' or ':' (for perf-like syntax);  followed by an
+     *   optional separator:
+     *
+     *     <pmu>/<event_name>[/]
+     *   OR
+     *     <pmu>:<event_name>[/]
+     *
+     *   Examples (both specify the same event):
+     *
+     *     cpu/cache-misses/
+     *     cpu:cache-misses
+     *
+     * */
+
+    enum EVENT_DESCRIPTION_REGEX_GROUPS
+    {
+        ED_WHOLE_MATCH,
+        ED_PMU,
+        ED_NAME,
+    };
+
+    parse_pmu_path(ev_name);
+
+    Log::debug() << "parsing event description: pmu='" << pmu_name_ << "', event='" << name_ << "'";
+
+    // read PMU type id
+    std::underlying_type<perf_type_id>::type type = read_file_or_else(pmu_path_ / "type", 0);
```

This `read_file_or_else` business is very strange. You're making a function, that reads a file, but ignores errors and returns a sentinel value instead. Then you check for the same sentinel value and throw an error.

---

Which alternative would you suggest? Copying the ifstream code from read_file_or_else every time or making some try_read_file function, which would also copy the code? 
I agree that this kinda alienates the function, but at the same time it avoids code duplication.

---

The issue I see here is that it feels like you're trying to import error handling from a different language into C++. `X_or_else` feels very rusty. A more idiomatic way to implement `read_file` would be to either return a `std::optional` or throw an exception.